File: Web_Stadium/Services/TournamentExcelService.cs
using ClosedXML.Excel;
using Microsoft.EntityFrameworkCore;
using System.Drawing;
using Web_Stadium.EFCore;

namespace Web_Stadium.Services
{
    /// <summary>
    /// Infrastructure concern: xuất Excel đối soát tài chính giải đấu
    /// Không thuộc domain — tách hoàn toàn
    /// </summary>
    public class TournamentExcelService
    {
        private readonly SanBongContext _context;
        public TournamentExcelService(SanBongContext context) => _context = context;

        // ══════════════════════════════════════════════════════════
        // Xuất file Excel 3 sheet: tổng hợp + thẻ phạt + kết quả
        // ══════════════════════════════════════════════════════════
        public async Task<(byte[] bytes, string fileName)> ExportDoiSoat(int giaiDauId)
        {
            var giai = await _context.GiaiDaus
                .Include(g => g.DoiBongs).ThenInclude(d => d.ThanhViens)
                .Include(g => g.TranDaus).ThenInclude(t => t.SuKiens)
                    .ThenInclude(s => s.ThanhVien)
                .Include(g => g.TranDaus).ThenInclude(t => t.DoiNha)
                .Include(g => g.TranDaus).ThenInclude(t => t.DoiKhach)
                .FirstOrDefaultAsync(g => g.Id == giaiDauId)
                ?? throw new KeyNotFoundException("Không tìm thấy giải đấu!");

            using var wb = new XLWorkbook();

            BuildSheet1_TongHop(wb, giai);
            BuildSheet2_ThePhat(wb, giai);
            BuildSheet3_KetQua(wb, giai);

            using var stream = new MemoryStream();
            wb.SaveAs(stream);

            var safeFileName = string.Concat(
                giai.TenGiai.Select(c => Path.GetInvalidFileNameChars().Contains(c) ? '_' : c));
            var fileName = $"DoiSoat_{safeFileName}_{DateTime.Now:yyyyMMdd}.xlsx";

            return (stream.ToArray(), fileName);
        }

        // ── Sheet 1: Tổng hợp từng đội ───────────────────────────
        private void BuildSheet1_TongHop(XLWorkbook wb, GiaiDau giai)
        {
            var ws = wb.AddWorksheet("Tổng hợp đội");

            // Title
            ws.Cell(1, 1).Value = $"GIẢI ĐẤU: {giai.TenGiai.ToUpper()}";
            ws.Cell(1, 1).Style.Font.Bold = true;
            ws.Cell(1, 1).Style.Font.FontSize = 14;
            ws.Cell(1, 1).Style.Font.FontColor = XLColor.FromArgb(14, 168, 106);
            ws.Range(1, 1, 1, 9).Merge();

            ws.Cell(2, 1).Value = $"Xuất lúc: {DateTime.Now:dd/MM/yyyy HH:mm}";
            ws.Cell(2, 1).Style.Font.Italic = true;
            ws.Range(2, 1, 2, 9).Merge();

            // Header
            var headers = new[] {
                "STT", "Tên đội", "Ký quỹ ban đầu",
                "Thẻ vàng", "Phạt vàng (đ)",
                "Thẻ đỏ",  "Phạt đỏ (đ)",
                "Tổng phạt", "Hoàn trả thực tế"
            };
            for (int c = 0; c < headers.Length; c++)
            {
                var cell = ws.Cell(4, c + 1);
                cell.Value = headers[c];
                cell.Style.Font.Bold = true;
                cell.Style.Fill.BackgroundColor = XLColor.FromArgb(14, 168, 106);
                cell.Style.Font.FontColor = XLColor.White;
                cell.Style.Alignment.Horizontal = XLAlignmentHorizontalValues.Center;
            }

            // Data
            int row = 5, stt = 1;
            decimal tongHoanTra = 0;

            foreach (var doi in giai.DoiBongs.Where(d => d.DaThanhToan))
            {
                var suKienDoi = giai.TranDaus
                    .SelectMany(t => t.SuKiens)
                    .Where(s => s.DoiId == doi.Id).ToList();

                var soVang = suKienDoi.Count(s => s.LoaiSuKien is "TheVang" or "TheVangLan2");
                var soDo = suKienDoi.Count(s => s.LoaiSuKien == "TheDo");
                var phatVang = soVang * giai.TienPhatTheVang;
                var phatDo = soDo * giai.TienPhatTheDo;
                var tongPhat = phatVang + phatDo;
                var hoanTra = Math.Max(0, giai.TienKyQuy - tongPhat);
                tongHoanTra += hoanTra;

                ws.Cell(row, 1).Value = stt++;
                ws.Cell(row, 2).Value = doi.TenDoi;
                ws.Cell(row, 3).Value = (double)giai.TienKyQuy;
                ws.Cell(row, 4).Value = soVang;
                ws.Cell(row, 5).Value = (double)phatVang;
                ws.Cell(row, 6).Value = soDo;
                ws.Cell(row, 7).Value = (double)phatDo;
                ws.Cell(row, 8).Value = (double)tongPhat;
                ws.Cell(row, 9).Value = (double)hoanTra;

                // Tô màu theo mức độ phạt
                var rowStyle = ws.Row(row).Style;
                if (tongPhat >= giai.TienKyQuy)
                    rowStyle.Fill.BackgroundColor = XLColor.FromArgb(255, 220, 220);
                else if (tongPhat > 0)
                    rowStyle.Fill.BackgroundColor = XLColor.FromArgb(255, 248, 220);

                row++;
            }

            // Total row
            ws.Cell(row, 2).Value = "TỔNG HOÀN TRẢ";
            ws.Cell(row, 9).Value = (double)tongHoanTra;
            ws.Range(row, 1, row, 9).Style.Font.Bold = true;
            ws.Range(row, 1, row, 9).Style.Fill.BackgroundColor = XLColor.FromArgb(220, 240, 230);

            ws.Columns().AdjustToContents();
            FormatCurrencyColumns(ws, new[] { 3, 5, 7, 8, 9 }, 4, row);
        }

        // ── Sheet 2: Chi tiết thẻ phạt ───────────────────────────
        private void BuildSheet2_ThePhat(XLWorkbook wb, GiaiDau giai)
        {
            var ws = wb.AddWorksheet("Chi tiết thẻ phạt");

            var headers = new[] { "Vòng", "Trận đấu", "Đội", "Cầu thủ", "Loại thẻ", "Phút", "Tiền phạt" };
            for (int c = 0; c < headers.Length; c++)
            {
                ws.Cell(1, c + 1).Value = headers[c];
                ws.Cell(1, c + 1).Style.Font.Bold = true;
                ws.Cell(1, c + 1).Style.Fill.BackgroundColor = XLColor.FromArgb(41, 98, 255);
                ws.Cell(1, c + 1).Style.Font.FontColor = XLColor.White;
            }

            int row = 2;
            foreach (var tran in giai.TranDaus.OrderBy(t => t.VongDau))
            {
                var theEvents = tran.SuKiens
                    .Where(s => s.LoaiSuKien is "TheVang" or "TheDo" or "TheVangLan2")
                    .ToList();

                foreach (var sk in theEvents)
                {
                    var tenDoi = giai.DoiBongs.FirstOrDefault(d => d.Id == sk.DoiId)?.TenDoi ?? "?";
                    var tienPhat = sk.LoaiSuKien == "TheDo"
                        ? giai.TienPhatTheDo : giai.TienPhatTheVang;

                    ws.Cell(row, 1).Value = $"Vòng {tran.VongDau}";
                    ws.Cell(row, 2).Value = $"{tran.DoiNha?.TenDoi} vs {tran.DoiKhach?.TenDoi}";
                    ws.Cell(row, 3).Value = tenDoi;
                    ws.Cell(row, 4).Value = sk.ThanhVien?.HoTen ?? "?";
                    ws.Cell(row, 5).Value = sk.LoaiSuKien;
                    ws.Cell(row, 6).Value = sk.Phut?.ToString() ?? "-";
                    ws.Cell(row, 7).Value = (double)tienPhat;

                    // Màu thẻ
                    ws.Cell(row, 5).Style.Fill.BackgroundColor =
                        sk.LoaiSuKien == "TheDo"
                            ? XLColor.FromArgb(255, 100, 100)
                            : XLColor.FromArgb(255, 220, 100);

                    row++;
                }
            }

            ws.Columns().AdjustToContents();
        }

        // ── Sheet 3: Kết quả toàn giải ───────────────────────────
        private void BuildSheet3_KetQua(XLWorkbook wb, GiaiDau giai)
        {
            var ws = wb.AddWorksheet("Kết quả trận đấu");

            var headers = new[] { "Vòng", "Loại", "Đội nhà", "Tỷ số", "Đội khách", "Trạng thái" };
            for (int c = 0; c < headers.Length; c++)
            {
                ws.Cell(1, c + 1).Value = headers[c];
                ws.Cell(1, c + 1).Style.Font.Bold = true;
                ws.Cell(1, c + 1).Style.Fill.BackgroundColor = XLColor.FromArgb(245, 158, 11);
                ws.Cell(1, c + 1).Style.Font.FontColor = XLColor.White;
            }

            int row = 2;
            foreach (var t in giai.TranDaus.OrderBy(t => t.VongDau).ThenBy(t => t.NgayThiDau))
            {
                var tyso = t.BanThangNha.HasValue
                    ? $"{t.BanThangNha} - {t.BanThangKhach}" : "Chưa đấu";

                ws.Cell(row, 1).Value = t.VongDau;
                ws.Cell(row, 2).Value = t.LoaiVong;
                ws.Cell(row, 3).Value = t.DoiNha?.TenDoi ?? "?";
                ws.Cell(row, 4).Value = tyso;
                ws.Cell(row, 5).Value = t.DoiKhach?.TenDoi ?? "?";
                ws.Cell(row, 6).Value = t.TrangThai;

                ws.Cell(row, 4).Style.Alignment.Horizontal = XLAlignmentHorizontalValues.Center;
                ws.Cell(row, 4).Style.Font.Bold = t.TrangThai == "Closed";

                row++;
            }

            ws.Columns().AdjustToContents();
        }

        // ── Format tiền tệ ───────────────────────────────────────
        private void FormatCurrencyColumns(
            IXLWorksheet ws, int[] cols, int startRow, int endRow)
        {
            foreach (var col in cols)
                ws.Range(startRow, col, endRow, col)
                  .Style.NumberFormat.Format = "#,##0";
        }
    }
}

File: tournament-service/src/main/java/com/pitchhub/tournament/service/TournamentExcelService.java
package com.pitchhub.tournament.service;

import com.pitchhub.tournament.model.DoiBong;
import com.pitchhub.tournament.model.GiaiDau;
import com.pitchhub.tournament.model.SuKienTran;
import com.pitchhub.tournament.model.TranDau;
import com.pitchhub.tournament.repository.GiaiDauRepository;
import org.apache.poi.ss.usermodel.*;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;
import org.springframework.stereotype.Service;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.math.BigDecimal;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.NoSuchElementException;

/**
 * Infrastructure concern: xuat Excel doi soat tai chinh giai dau.
 * Port tu TournamentExcelService.cs — dung Apache POI thay ClosedXML.
 */
@Service
public class TournamentExcelService {

    private final GiaiDauRepository giaiDauRepository;

    public TournamentExcelService(GiaiDauRepository giaiDauRepository) {
        this.giaiDauRepository = giaiDauRepository;
    }

    public record ExcelFile(byte[] bytes, String fileName) {
    }

    public ExcelFile exportDoiSoat(Integer giaiDauId) {
        GiaiDau giai = giaiDauRepository.findDetailById(giaiDauId)
                .orElseThrow(() -> new NoSuchElementException("Khong tim thay giai dau!"));

        try (XSSFWorkbook wb = new XSSFWorkbook()) {
            buildSheet1TongHop(wb, giai);
            buildSheet2ThePhat(wb, giai);
            buildSheet3KetQua(wb, giai);

            try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
                wb.write(out);
                String safeFileName = giai.getTenGiai().replaceAll("[\\\\/:*?\"<>|]", "_");
                String fileName = "DoiSoat_" + safeFileName + "_"
                        + java.time.LocalDate.now().format(DateTimeFormatter.ofPattern("yyyyMMdd")) + ".xlsx";
                return new ExcelFile(out.toByteArray(), fileName);
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    // ── Sheet 1: Tong hop tung doi ───────────────────────────
    private void buildSheet1TongHop(XSSFWorkbook wb, GiaiDau giai) {
        Sheet ws = wb.createSheet("Tong hop doi");

        CellStyle titleStyle = wb.createCellStyle();
        Font titleFont = wb.createFont();
        titleFont.setBold(true);
        titleFont.setFontHeightInPoints((short) 14);
        titleFont.setColor(IndexedColors.DARK_GREEN.getIndex());
        titleStyle.setFont(titleFont);

        Row titleRow = ws.createRow(0);
        Cell titleCell = titleRow.createCell(0);
        titleCell.setCellValue("GIAI DAU: " + giai.getTenGiai().toUpperCase());
        titleCell.setCellStyle(titleStyle);

        CellStyle headerStyle = headerStyle(wb, IndexedColors.DARK_GREEN);
        String[] headers = {"STT", "Ten doi", "Ky quy ban dau", "The vang", "Phat vang (d)",
                "The do", "Phat do (d)", "Tong phat", "Hoan tra thuc te"};
        Row headerRow = ws.createRow(3);
        for (int c = 0; c < headers.length; c++) {
            Cell cell = headerRow.createCell(c);
            cell.setCellValue(headers[c]);
            cell.setCellStyle(headerStyle);
        }

        int rowIdx = 4, stt = 1;
        BigDecimal tongHoanTra = BigDecimal.ZERO;

        for (DoiBong doi : giai.getDoiBongs()) {
            if (!Boolean.TRUE.equals(doi.getDaThanhToan())) continue;

            List<SuKienTran> suKienDoi = giai.getTranDaus().stream()
                    .flatMap(t -> t.getSuKiens().stream())
                    .filter(s -> doi.getId().equals(s.getDoiId()))
                    .toList();

            long soVang = suKienDoi.stream().filter(s -> "TheVang".equals(s.getLoaiSuKien()) || "TheVangLan2".equals(s.getLoaiSuKien())).count();
            long soDo = suKienDoi.stream().filter(s -> "TheDo".equals(s.getLoaiSuKien())).count();
            BigDecimal phatVang = giai.getTienPhatTheVang().multiply(BigDecimal.valueOf(soVang));
            BigDecimal phatDo = giai.getTienPhatTheDo().multiply(BigDecimal.valueOf(soDo));
            BigDecimal tongPhat = phatVang.add(phatDo);
            BigDecimal hoanTra = giai.getTienKyQuy().subtract(tongPhat).max(BigDecimal.ZERO);
            tongHoanTra = tongHoanTra.add(hoanTra);

            Row row = ws.createRow(rowIdx);
            row.createCell(0).setCellValue(stt++);
            row.createCell(1).setCellValue(doi.getTenDoi());
            row.createCell(2).setCellValue(giai.getTienKyQuy().doubleValue());
            row.createCell(3).setCellValue(soVang);
            row.createCell(4).setCellValue(phatVang.doubleValue());
            row.createCell(5).setCellValue(soDo);
            row.createCell(6).setCellValue(phatDo.doubleValue());
            row.createCell(7).setCellValue(tongPhat.doubleValue());
            row.createCell(8).setCellValue(hoanTra.doubleValue());

            if (tongPhat.compareTo(giai.getTienKyQuy()) >= 0) {
                toMauDong(wb, row, IndexedColors.ROSE);
            } else if (tongPhat.compareTo(BigDecimal.ZERO) > 0) {
                toMauDong(wb, row, IndexedColors.LIGHT_YELLOW);
            }

            rowIdx++;
        }

        CellStyle boldStyle = wb.createCellStyle();
        Font boldFont = wb.createFont();
        boldFont.setBold(true);
        boldStyle.setFont(boldFont);

        Row totalRow = ws.createRow(rowIdx);
        Cell tongLabel = totalRow.createCell(1);
        tongLabel.setCellValue("TONG HOAN TRA");
        tongLabel.setCellStyle(boldStyle);
        Cell tongValue = totalRow.createCell(8);
        tongValue.setCellValue(tongHoanTra.doubleValue());
        tongValue.setCellStyle(boldStyle);

        for (int c = 0; c < headers.length; c++) ws.autoSizeColumn(c);
    }

    // ── Sheet 2: Chi tiet the phat ───────────────────────────
    private void buildSheet2ThePhat(XSSFWorkbook wb, GiaiDau giai) {
        Sheet ws = wb.createSheet("Chi tiet the phat");

        CellStyle headerStyle = headerStyle(wb, IndexedColors.BLUE);
        String[] headers = {"Vong", "Tran dau", "Doi", "Cau thu", "Loai the", "Phut", "Tien phat"};
        Row headerRow = ws.createRow(0);
        for (int c = 0; c < headers.length; c++) {
            Cell cell = headerRow.createCell(c);
            cell.setCellValue(headers[c]);
            cell.setCellStyle(headerStyle);
        }

        int rowIdx = 1;
        List<TranDau> tranDaus = giai.getTranDaus().stream()
                .sorted((a, b) -> Integer.compare(a.getVongDau(), b.getVongDau())).toList();

        for (TranDau tran : tranDaus) {
            List<SuKienTran> theEvents = tran.getSuKiens().stream()
                    .filter(s -> "TheVang".equals(s.getLoaiSuKien()) || "TheDo".equals(s.getLoaiSuKien())
                            || "TheVangLan2".equals(s.getLoaiSuKien()))
                    .toList();

            for (SuKienTran sk : theEvents) {
                String tenDoi = giai.getDoiBongs().stream()
                        .filter(d -> d.getId().equals(sk.getDoiId())).findFirst()
                        .map(DoiBong::getTenDoi).orElse("?");
                BigDecimal tienPhat = "TheDo".equals(sk.getLoaiSuKien()) ? giai.getTienPhatTheDo() : giai.getTienPhatTheVang();

                Row row = ws.createRow(rowIdx++);
                row.createCell(0).setCellValue("Vong " + tran.getVongDau());
                row.createCell(1).setCellValue(tenDoiAn(tran.getDoiNha()) + " vs " + tenDoiAn(tran.getDoiKhach()));
                row.createCell(2).setCellValue(tenDoi);
                row.createCell(3).setCellValue(sk.getThanhVien() != null ? sk.getThanhVien().getHoTen() : "?");
                row.createCell(4).setCellValue(sk.getLoaiSuKien());
                row.createCell(5).setCellValue(sk.getPhut() != null ? sk.getPhut().toString() : "-");
                row.createCell(6).setCellValue(tienPhat.doubleValue());
            }
        }

        for (int c = 0; c < headers.length; c++) ws.autoSizeColumn(c);
    }

    // ── Sheet 3: Ket qua toan giai ───────────────────────────
    private void buildSheet3KetQua(XSSFWorkbook wb, GiaiDau giai) {
        Sheet ws = wb.createSheet("Ket qua tran dau");

        CellStyle headerStyle = headerStyle(wb, IndexedColors.ORANGE);
        String[] headers = {"Vong", "Loai", "Doi nha", "Ty so", "Doi khach", "Trang thai"};
        Row headerRow = ws.createRow(0);
        for (int c = 0; c < headers.length; c++) {
            Cell cell = headerRow.createCell(c);
            cell.setCellValue(headers[c]);
            cell.setCellStyle(headerStyle);
        }

        int rowIdx = 1;
        List<TranDau> tranDaus = giai.getTranDaus().stream()
                .sorted((a, b) -> {
                    int cmp = Integer.compare(a.getVongDau(), b.getVongDau());
                    return cmp != 0 ? cmp : a.getNgayThiDau().compareTo(b.getNgayThiDau());
                }).toList();

        for (TranDau t : tranDaus) {
            String tyso = t.getBanThangNha() != null ? t.getBanThangNha() + " - " + t.getBanThangKhach() : "Chua dau";

            Row row = ws.createRow(rowIdx++);
            row.createCell(0).setCellValue(t.getVongDau());
            row.createCell(1).setCellValue(t.getLoaiVong());
            row.createCell(2).setCellValue(tenDoiAn(t.getDoiNha()));
            row.createCell(3).setCellValue(tyso);
            row.createCell(4).setCellValue(tenDoiAn(t.getDoiKhach()));
            row.createCell(5).setCellValue(t.getTrangThai());
        }

        for (int c = 0; c < headers.length; c++) ws.autoSizeColumn(c);
    }

    private String tenDoiAn(DoiBong doi) {
        return doi != null ? doi.getTenDoi() : "?";
    }

    private CellStyle headerStyle(XSSFWorkbook wb, IndexedColors color) {
        CellStyle style = wb.createCellStyle();
        Font font = wb.createFont();
        font.setBold(true);
        font.setColor(IndexedColors.WHITE.getIndex());
        style.setFont(font);
        style.setFillForegroundColor(color.getIndex());
        style.setFillPattern(FillPatternType.SOLID_FOREGROUND);
        style.setAlignment(HorizontalAlignment.CENTER);
        return style;
    }

    private void toMauDong(XSSFWorkbook wb, Row row, IndexedColors color) {
        CellStyle style = wb.createCellStyle();
        style.setFillForegroundColor(color.getIndex());
        style.setFillPattern(FillPatternType.SOLID_FOREGROUND);
        row.forEach(cell -> cell.setCellStyle(style));
    }
}
